### ISSUE TITLE
2.16 deserialization changes have broken support for the scala.util.Either class

### ISSUE THREAD

**Issue by pjfanning** [FasterXML/jackson-databind#4147]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

The jackson-module-scala tests haven't run since mid August when they worked. Something has changed in jackson-databind deserialization code that has broken 4 tests in jackson-module-scala - all related to deserializing scala.util.Either. 

Please don't release jackson-databind 2.16 until this can be investigated. Maybe the EitherDeserializer depends on some behaviour it shouldn't have relied on but so far this looks like a jackson-databind regression.

The issue also now affects master branch but not the 2.15 branch.

https://github.com/FasterXML/jackson-module-scala/issues/649

### Version Information

2.16 latest snapshots and 3.0.0 latest snapshots

**Comment by pjfanning** [comment]
I've been getting some variable results but so far I seem to see this.
* 2.16.0-20230912.165553-141 works
* 2.16.0-20230920.234330-142 fails

So possibly a change that went into jackson-databind in mid September

**Comment by pjfanning** [comment]
I don't know for definite yet but I suspect that this issue is related to https://github.com/FasterXML/jackson-databind/pull/4122 - the test failures in jackson-module-scala relate to deserialization where wildcard types are used.

**Comment by pjfanning** [comment]
The issue seems to be that Either has 2 type bindings `Either<L, R>` but jackson-databind only fills in one type binding of the other one is wildcard. I have added a hack to jackson-module-scala 2.16 to check the binding name so I can try to work out if the binding relates to L or R. In jackson-databind, there are 2 bindings provided on the ReferenceType regardless of whether one of them is a wildcard or not.

**Comment by cowtowncoder** [comment]
Ok, I went ahead and reverted #4122 so that this issue (and new tests via #4149) pass.
Hoping to tackle #4122 again for 2.17 as per PR @yawkat submitted (f.ex), but due to timing seems like time to do that is not now.

Thank you @pjfanning for surfacing these issues: I know you raised them before my merging of aborted fix. But I think the process worked ok here nonetheless -- we found risk warranted revert in this case and doing that hopefully makes 2.16.0 a safer upgrade for our user base.
